tools/base/fpgainfo/board.c

```diff
@@ -0,0 +1,440 @@
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+#ifndef __USE_GNU
+#define __USE_GNU
+#endif
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <getopt.h>
+#include "fpgainfo.h"
+#include <opae/fpga.h>
+#include <wchar.h>
+#include <dirent.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <dlfcn.h>
+#include <pthread.h>
+
+#include "safe_string/safe_string.h"
+
+#include "board.h"
+
+
+static pthread_mutex_t board_plugin_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
+
+// Board plug-in table
+static platform_data platform_data_table[] = {
+	{ 0x8086, 0x09c4, "libboard_rc.so", NULL },
+	{ 0x8086, 0x09c5, "libboard_rc.so", NULL },
+	{ 0,      0,          NULL, NULL },
+};
+
+
+fpga_result load_board_plugin(fpga_token token, void** dl_handle)
+{
+	fpga_result res                = FPGA_OK;
+	fpga_properties props          = NULL;
+	uint16_t vendor_id             = 0;
+	uint16_t device_id             = 0;
+	int i                          = 0;
+
+	if (token == NULL || dl_handle == NULL) {
+		OPAE_ERR("Invalid input parameter");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaGetProperties(token, &props);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get properties\n");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaPropertiesGetDeviceID(props, &device_id);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get device ID\n");
+		goto err_out_destroy;
+	}
+
+	res = fpgaPropertiesGetVendorID(props, &vendor_id);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get vendor ID\n");
+		goto err_out_destroy;
+	}
+
+	if (pthread_mutex_lock(&board_plugin_lock) != 0 ) {
+		OPAE_ERR("pthread mutex lock failed \n");
+
+		if (props)
+			fpgaDestroyProperties(&props);
+		return FPGA_EXCEPTION;
+	}
+
+	for (i = 0; platform_data_table[i].board_plugin; ++i) {
+
+		if (platform_data_table[i].device_id == device_id &&
+			platform_data_table[i].vendor_id == vendor_id) {
+
+			// Loaded lib or found
+			if (platform_data_table[i].dl_handle) {
+				*dl_handle = platform_data_table[i].dl_handle;
+				goto err_out_destroy;
+			}
+
+			platform_data_table[i].dl_handle = dlopen(platform_data_table[i].board_plugin, RTLD_LAZY | RTLD_LOCAL);
+			if (!platform_data_table[i].dl_handle) {
+				char *err = dlerror();
+				OPAE_ERR("Failed to load \"%s\" %s", platform_data_table[i].board_plugin, err ? err : "");
+				goto err_out_destroy;
+			} else {
+				*dl_handle = platform_data_table[i].dl_handle;
+				goto err_out_destroy;
+			}
+		} //end if
+
+	} // end for
+
+
+
+err_out_destroy:
+	if (props)
+		fpgaDestroyProperties(&props);

```

Comment:
Are we not checking the return code of this?

Reply:
fxied